### ISSUE TITLE
resource_openstack_compute_instance_v2 doesn't support project_id or tenant_id property

### ISSUE THREAD

**Issue by t928948** [terraform-provider-openstack/terraform-provider-openstack#1294]
Hi there,
I've noticed that current implementation of openstack_compute_instance and openstack_blockstorage_volume does not have tenant_id property.

### Use case when it is needed:
We have service account for automation which has admin rights: service1
1. initialize openstack provider with service1 in admin project.
2. Create openstack project "project2"
3. Create create network in project2
4. create Server and Volume in project2.

### Expected Behavior
Project, network, server and volume are created
### Actual Behavior
Project, network are created.
Server and Volume can't be created under project2

**Comment by wsidl** [comment]
Just noticed this as well. How you fix this is configure multiple providers for your project.

```hcl
# Auth in clouds.yaml
provider "openstack" {
  cloud = "admin"     
}

provider "openstack" {
  alias = "project"
  cloud = "project"
}
```

Then when you define your resource or data blocks, you include the `provider` meta property to define which provider it should be targeting.

```hcl
resource "openstack_compute_instance_v2" "nats" {
  # assumes `provider = openstack`
  name = "nats"
  ...
}

resource "openstack_compute_instance_v2" "proxy" {
  provider = openstack.project
  name     = "proxy"
  ...
}
```

**Comment by jbe-dw** [comment]
This is really annoying, we have a code that setup all our openstack infrastructure content (tenants, networks, subnets, default rules) and we want to automatically deploy some service instances in each tenant, but we can't.

While a lot of networking modules support it, we are stopped by these two modules.

@wsidl We have a dynamic code were all tenants (60+) are defined in a yaml file. Statically describing a provider config for each tenant is clearly not the way to do.

This also means that if you have any project that massively want to deploy instances over several tenants, we have to stick to the old terraform philosphy of 1 res = 1 block = unamainainable large chunks of redundant code.

**Comment by wsidl** [comment]
@jbe-dw not quite sure what the issue is here. Kind of sounds like you're duplicating the resources across tenants/projects? If there's anything you can do to ease the issue using modules?

**Comment by jbe-dw** [comment]
Yes we provide a tenant with a network and a subnet, assign remote users to the tenant, register everything in the cmdb, provide a consul keystore for users to store their tfstate concerning the tenant. We also would like to provide services served by prebuilt instances in each tenant.

To do so, we have a private module that build instances (network port, cinder volumes, nova instance). We call the module in a for_each loop. We can't call dynamically a provider config alias because we use for_each and it seems not supported but even if it could, it sounds like we should at least write a terraform provider config for each tenant, which is not ... dynamic.

If we could pass a tenant_id parameter to these two resources, we would be able to add such a parameter to the module.

What we are going to do is to create the instances in a management tenant, which have a shared access to the other tenant's networks (maybe it is better because they will be hidden from the tenant they serve).

**Comment by t928948** [comment]
Our original use case for described issue is provisioning anything for a new tenant. I.e. we can't configure new provider, because the project for that provider is just about to be created.

I.e. simple example which shall be possible:

```
provider "openstack" {
  user_name   = "admin"
  tenant_name = "admin"
  password    = "pwd"
  auth_url    = "http://myauthurl:5000/v2.0"
  region      = "RegionOne"
}

resource "openstack_identity_project_v3" "project2" {
  name        = project2
  description = "project2"
}

resource "openstack_networking_network_v2" "net1" {
  name                  = "network"
  admin_state_up        = "true"
  tenant_id  = openstack_identity_project_v3.project2.id    <========= this is possible
}

resource "openstack_blockstorage_volume_v2" "myvol" {
  name = "myvol"
  size = 1

  tenant_id = openstack_identity_project_v3.project2.id     <========= this is NOT possible
}

resource "openstack_compute_instance_v2" "myinstance" {
  name            = "myinstance"
  image_id        = "ad091b52-742f-469e-8f3c-fd81cadf0743"
  flavor_id       = "3"
  key_pair        = "my_key_pair_name"
  security_groups = ["default"]

  network {
    uuid= openstack_networking_network_v2.net1.id
  }

  tenant_id = openstack_identity_project_v3.project2.id  <========= this is NOT possible

}

resource "openstack_compute_volume_attach_v2" "attached" {
  instance_id = "${openstack_compute_instance_v2.myinstance.id}"
  volume_id   = "${openstack_blockstorage_volume_v2.myvol.id}"
  
  tenant_id = openstack_identity_project_v3.project2.id  <========= this is NOT possible
}

```

**Comment by wsidl** [comment]
@t928948 , I managed to do this by setting up the cloud file with the project details to be created. Then I would set up 2 providers (one for `admin` and the other for the project) and then set the `provider` to the alias.

Something like:

```hcl
provider "openstack" {
  user_name = "admin"
  ...
}
provider "openstack" {
  alias       = "project2"
  tenant_name = "project2"
  ...
}

...

resource "openstack_networking_network_v2" "net1" {
  provider = openstack.project2
  name     = "network"
  ...
}

**Comment by jbe-dw** [comment]
The limitation also affects ssh_keypair creation. What we do:

- Describe all our tenants/networks in a YAML file, so any modification to the cloud is just about changing the YAML content.
- Have terraform code using modules and a for_each loop over tenants loaded from yaml to create everything in openstack, but in our cmdb, or create dns domains for each tenant

We could work the problem around for the first batch of servers by using alternative provider configuration, but we had to manually create one tenant configuration and call the server module with this provider. All the instances are hosted on the same tenant but they are connected to other tenants networks.

We would like to create ssh keypairs for each tenant member who have it in LDAP, but the keypairs are tenant scoped. So if we would like to achieve this with your proposal, we would have to write terraform code to describe a provider for each tenant (~100), which is the opposite way of what we are doing now (using a single source of data, no more "per tenant terraform code" - the tf code has been reduced by 75%).

Would it be that hard and a lot of work to add tenant_id support in those resources, like it is possible in the network resources ? 

2y ago TF was just about describing everything, but it doesn't scale (copy paste code is subject to many mistakes), so they finally add support for more looping logic. The provider thing is a last blocking point since the values inside must be known before applying anything. For a cloud admin, just changing the tenant name in the client before applying is enough. This is why it could make our life much easier to add this parameter in resources.

Could you at least give us an answer about this request ? (no because ... ?)

Regards.

**Comment by nikParasyr** [comment]
From a quick look i had on api docs:

Defining `project_id/tenant_id` is not allowed on nova api level based on docs [here](https://docs.openstack.org/api-ref/compute/?expanded=create-server-detail#create-server). In comparison neutron api for creating networks allows that as seen [here](https://docs.openstack.org/api-ref/network/v2/index.html?expanded=create-network-detail#networks).

Therefore this does **not** seem to be the case where it is possible on openstack/api level and we have not implemented on the provider. As long as it is not supported on an api level then i'd say its not feasible to implement anything on the provider.

This is just my input/opinion from a quick look i had. I might have missed something though.

**Comment by jbe-dw** [comment]
Yes but if it is not an option as an API call parameter, it could be possible to overload the client config if the resource option is active ? The placement is based on the token scope requested by the client.

The API option is available in cinder ([doc here](https://docs.openstack.org/api-ref/block-storage/v3/?expanded=create-a-volume-detail).

**Comment by Banshee1221** [comment]
I'm adding my voice to this..

I have an R&D cloud in which I create labs for students to do training on. What we want to do is be able to define a variable file containing information about each student team, then have terraform create a project/user for each team and create the standardised resources on respective projects.

**Comment by nevesm** [comment]
Waking up this discussion, i am facing this issue right now. I need to create the project using terraform, but can't create a `openstack_compute_instance_v2` inside that because this resource doesn't have `tenant_id` or `tenant_name` properties...

What are you guys doing as a workaround on this case?

**Comment by vivienbcr** [comment]
Exactly same problem as @nevesm did someone find workaround ?

**Comment by nevesm** [comment]
I am using this workaround:

1 repository to create tenants, and then using this tenant created on the provider settings of other repositories, i have a structure with multiple clients and i create 1 repo per client to create the resources.

example:

Repository: `openstack-tenant-administration-infra` - This repo has a default tenant in the provider.tf file, which is `admin`, and creates multiple tenants based on a variable:

`provider.tf`
```
terraform {
  required_providers {
    openstack = {
      source = "terraform-provider-openstack/openstack"
      version = "~> 3.0.0"
    }
  }
}

provider "openstack" {
  auth_url                      = "{{AUTH_URL}}"
  application_credential_id     = "{{APPLICATION_CREDENTIAL_ID}}"
  application_credential_secret = "{{APPLICATION_CREDENTIAL_SECRET}}"
  application_credential_name   = "{{APPLICATION_CREDENTIAL_NAME}}"
  region                        = "{{REGION}}"
  tenant_name                   = "admin"
}
```

`identity.tf`
```
 resource "openstack_identity_project_v3" "main" {
  for_each = {
    for tenant, environments in var.tenants : 
    tenant => environments[var.environment] if contains(keys(environments), var.environment)
  }

  name        = each.value
  description = "Created by Terraform"
}

```

`variables.tf`
```
variable "tenants" {
  description = "The list of tenants to create"
  default = {
    client1 = {
      prod = "client1-prod"
    }
    client2 = {
      prod = "client2-prod"
      qa     = "client2-qa"
    }
  }
}
```

Repository: `client2-infra` - I set the `tenant_name` in provider.tf and create the resources without passing any provider argument...

`provider.tf`
```
terraform {
  required_providers {
    openstack = {
      source  = "terraform-provider-openstack/openstack"
      version = "~> 3.0.0"
    }
  }
}

provider "openstack" {
  user_name   = "{{USER_NAME}}"
  password    = "{{PASSWORD}}"
  auth_url    = "{{AUTH_URL}}"
  region      = "{{REGION}}"
  tenant_name = "${var.tenant}-${var.environment}"
}
```

`variables.tf`
```
variable "environment" {
  type        = string
  description = "The environment for the deployment, this is populated by atlantis (prod | qa | dev)"
}

variable "project" {
  type        = string
  description = "The project name"
  default     = "client2-infra"
}

variable "tenant" {
  type        = string
  description = "The tenant name"
  default     = "client2"
}
```
